chapter2.adoc

```diff
@@ -1,214 +1,207 @@
 
-==== Submachine Mode Trace Control 
-The sec_check[i] signal for hart i in submachine mode is determined by the `sdetrcalw` field of CSR <<Sdseccsr, msdcfg>> within hart i, alongside mtrcen[i]. When the logical-OR of `sdetrcalw` and mtrcen[i] is 1, the sec_check[i] signal is cleared while the hart runs in submachine mode.
+[sdtrcctl]
+==== Supervisor Domain Trace Control 
+The Smsdetrc extension introduces `sdetrcalw` field (bit 8) in CSR <<Sdseccsr,msdcfg>> within hart i. The sec_check[i] signal for hart i in supervisor domain is determined by the `sdetrcalw` field and mtrcen[i]. When the logical-OR of `sdetrcalw` and mtrcen[i] is 1, the sec_check[i] signal is cleared while the hart runs in supervisor domain.
 
 When both `sdetrcalw` and mtrcen[i] are set to 0, the sec_check[i] signal cannot be cleared at all.
 
 [[trcctl]]
 [options="header"]
-.The status of the sec_check[i] sideband signal across privilege levels
+.Status of the sec_check[i] sideband signal across privilege levels
 |===========================================================
-| mtrcen| sdetrcalw| Machine mode | Submachine mode
+| mtrcen| sdetrcalw| M-mode | Supervisor domain
 | 1     | x     | sec_check[i] = 0   | sec_check[i] = 0     
 | 0     | 1     | sec_check[i] = 1   | sec_check[i] = 0     
 | 0     | 0     | sec_check[i] = 1   | sec_check[i] = 1     
 |===========================================================
 
 [NOTE]
-
 The sec_check signal serves as an additional signal for the trace module, indicating that trace output is prohibited due to security controls. Functionally, sec_check behaves identically to the halted signal. Both sec_check and halted signals cannot be active simultaneously. Reserved for future applications, the combined state of [sec_check, halted] as 0b11 remains unutilized. In cases where a trace module lacks support for the sec_check signal, the hart may alternatively toggle the halted signal to restrict trace output.
 
-=== Trigger 
+=== Trigger (Sdtrig)
 
-The trigger configured to enter Debug Mode is checked by Sdsec extension. The trigger can fire or match in privilege modes outlined in <<dbgpriv>>. 
+Triggers configured to enter Debug Mode can only fire or match when external debug is allowed, as outlined in <<dbgpriv>>. A trigger enabled for a privilege level higher than debug allowed privilege is not accessible by an external debugger. When this trigger is selected using `tselect`, it always reads as 0, and any writes to it are ignored.
 
-The extension requires that all pending triggers intending to enter Debug Mode must match or fire before any hart mode switch to prevent privilege escalation.
+[NOTE]
+Implementations shall make sure that pending triggers intending to enter Debug Mode match or fire before the privilege mode switch, thus the hart will never enter Debug Mode in higher privilege mode with debug disallowed  
```

Comment:
I agree that this statement is required. There might be pathological cases like the following: 
- mcontrol.timing=1 (after)
- Load matches the trigger in S-mode
- After that interrupt comes and then PC goes to M-mode interrupt handler in the next instruction. 
- According to RISC-V Debug Spec, DPC is set to the next instruction for timing=1 triggers, which is the trap handler in M-mode.

Reply:
This is really an issue. And the same case applies to single step when an interrupt occurs after instruction retirement. Since the dpc could be modified by the debugger, the debugger could deliberately suspend the trap handling by changing dpc to some other address. It could serve as some kind of timing attack. 

I propose to enforce that the interrupt should be handled before enter debug mode in case of trigger (after execution) or single step, what do you think of it?

Reply:
I assume hardware shall check if M-mode is allowed to debug then decide whether to update dpc i.e. only update dpc if debug is allowed.

I suggest add notes to highlight that to implementation

Reply:
Please note that single step with interrupt scenario can be mitigated by setting stepie=0. It is a fair assumption that if mdbgen=0, stepie is configured to zero by M-mode SW. It is also good that we restrict stepie configuration in this spec accordingly. We can add this restriction as a note to recommendation to the SW for single step. 
Regardless of single step problem, trigger with timing=1 still exists and your proposal makes sense to me add it as note.

Reply:
I am going to add the following note to elaborate on the case.
>Implementations must ensure that pending triggers intending to enter Debug Mode match or fire only when the hart is in a state where debug is allowed. For example, if an interrupt traps the hart to a debug-disallowed privilege mode, the trigger can only take effect either before the privilege is updated and control flow is transferred to the trap handler, or after the interrupt is completely handled and returns from the trap handler. The implementation must prevent Debug Mode from being entered in an intermediate state where privilege is changed or the PC is updated. This also applies to scenarios where a trigger is configured to enter Debug Mode before instruction execution and an interrupt occurs simultaneously.

Regarding the single step, the current definition only allows M-mode debugger to modify stepie and the default value is 0x0. That is, if we have a S-mode debugger, the interrupt traps to machine mode will unintentionally suppressed. I am thinking about making stepie always configurable in Debug Mode but only effective to interrupt delegated to debug-allowed privilege. The interrupt belongs to debug-disallowed privilege will fire regardless of stepie. What do you think of it?

Reply:
> Regarding the single step, the current definition only allows M-mode debugger to modify stepie and the default value is 0x0. That is, if we have a S-mode debugger, the interrupt traps to machine mode will unintentionally suppressed. I am thinking about making stepie always configurable in Debug Mode but only effective to interrupt delegated to debug-allowed privilege. The interrupt belongs to debug-disallowed privilege will fire regardless of stepie. What do you think of it?

Two options: 
1) I agree that the default state of stepie=0 is not convenient for mdbgen=0 case, but it is still under M-mode SW control. Securely implemented M-mode SW can enable it (stepie=1). 
2) Your proposal looks safe, but it adds HW complexity. 

I am fine with both of the options.

Reply:
Actually dcsr is actually only accessible in Debug Mode, thus option 1 might just not work...

Reply:
You are right. The option 1 would not work.